// adaptivecards-designer/lib/base-tree-item.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseTreeItem = void 0;
var draggable_element_1 = require("./draggable-element");
var adaptivecards_controls_1 = require("adaptivecards-controls");
var uuid_1 = require("uuid");
var BaseTreeItem = /** @class */ (function (_super) {
    __extends(BaseTreeItem, _super);
    function BaseTreeItem(isExpanded) {
        if (isExpanded === void 0) { isExpanded = true; }
        var _this = _super.call(this) || this;
        _this._isExpanded = true;
        _this._isSelected = false;
        _this._level = 0;
        _this._isExpanded = isExpanded;
        return _this;
    }
    BaseTreeItem.prototype.needsToBeScrolled = function () {
        // detect the case where the treeViewHost's parent isn't currently
        // overflowed. if it *is* overflowed (has scrollbar), then scrolling is
        // appropriate. otherwise, we might be in a small-screen situation
        // (single column rendering) where scrolling is undesireable
        var treeViewHostParent = this._rootElement.closest(".acd-treeView-host").parentElement;
        return (treeViewHostParent && treeViewHostParent.scrollHeight > treeViewHostParent.clientHeight);
    };
    BaseTreeItem.prototype.setIsSelected = function (value, scrollIntoView) {
        if (this._isSelected !== value) {
            this._isSelected = value;
            if (this._isSelected) {
                this._treeItemElement.classList.add("selected");
            }
            else {
                this._treeItemElement.classList.remove("selected");
            }
            this._rootElement.setAttribute("aria-current", this._isSelected.toString());
            this.selectedChanged(scrollIntoView);
        }
    };
    BaseTreeItem.prototype.click = function (e) {
        _super.prototype.click.call(this, e);
        this.setIsSelected(true, false);
    };
    BaseTreeItem.prototype.getIconClass = function () {
        return null;
    };
    BaseTreeItem.prototype.getAdditionalText = function () {
        return null;
    };
    BaseTreeItem.prototype.getAdditionalTextClass = function () {
        return "acd-tree-item-additionalText";
    };
    BaseTreeItem.prototype.getIndentationLevelIncrement = function () {
        return 8;
    };
    BaseTreeItem.prototype.getDragSourceElement = function () {
        return this._treeItemElement;
    };
    BaseTreeItem.prototype.selectedChanged = function (scrollIntoView) {
        if (this.isSelected && scrollIntoView && this.needsToBeScrolled()) {
            this._rootElement.scrollIntoView();
        }
        if (this.onSelectedChange) {
            this.onSelectedChange(this);
        }
    };
    BaseTreeItem.prototype.itemInvoked = function () {
        if (this.onItemInvoked) {
            this.onItemInvoked(this);
        }
    };
    BaseTreeItem.prototype.getExpandCollapseAriaText = function () {
        return (this._isExpanded ? "collapse" : "expand") + " " + this.getLabelText() + " element";
    };
    // collapse node or navigate to parent, if present
    BaseTreeItem.prototype.handleKeyboardLeftNav = function () {
        if (this._isExpanded && this.getChildCount() > 0) {
            // if we're an expanded node that has children, collapse the node.
            this._isExpanded = false;
            this.updateLayout();
            this._rootElement.tabIndex = 0;
            this._rootElement.focus();
        }
        else {
            // otherwise, we need to find the nearest ancestor treeitem or the root of the tree
            var nearestTreeItem = this._rootElement.parentElement.closest("[role='treeitem'], [role='tree']");
            var roleAttr = nearestTreeItem === null || nearestTreeItem === void 0 ? void 0 : nearestTreeItem.getAttribute("role");
            if (roleAttr === "treeitem") {
                nearestTreeItem.tabIndex = 0;
                nearestTreeItem.focus();
                this._rootElement.tabIndex = -1;
            }
            else if (roleAttr == "tree") {
                // only element higher up is the root DOM element of the tree, meaning `this`
                // represents the root node of the tree. set tabIndex so keyboard users can
                // tab back in.
                this._rootElement.tabIndex = 0;
                this._rootElement.focus();
            }
        }
    };
    // expand child nodes or navigate to first child, if present
    BaseTreeItem.prototype.handleKeyboardRightNav = function () {
        if (this.getChildCount() > 0) {
            if (this._isExpanded) {
                var childToFocus = this._childContainerElement.firstElementChild;
                childToFocus.tabIndex = 0;
                childToFocus.focus();
                this._rootElement.tabIndex = -1;
            }
            else {
                this._isExpanded = true;
                this.updateLayout();
            }
        }
    };
    BaseTreeItem.prototype.internalRender = function () {
        var _this = this;
        var hasChildren = (this.getChildCount() > 0);
        var labelId = (0, uuid_1.v4)();
        this._rootElement = document.createElement("li");
        this._rootElement.style.paddingLeft = "0";
        this._rootElement.style.listStyleType = "none";
        this._rootElement.setAttribute("aria-labelledby", labelId);
        this._rootElement.setAttribute("aria-level", this._level.toString());
        this._rootElement.setAttribute("aria-current", this._isSelected.toString());
        this._rootElement.tabIndex = -1;
        if (hasChildren) {
            this._rootElement.setAttribute("aria-expanded", this._isExpanded.toString());
        }
        this._rootElement.setAttribute("role", "treeitem");
        this._rootElement.onclick = function () { _this.click; };
        this._rootElement.onfocus = function () { _this.setIsSelected(true, false); };
        this._rootElement.onkeydown = function (e) {
            var handled = true;
            switch (e.key) {
                case adaptivecards_controls_1.Constants.keys.enter: {
                    _this.itemInvoked();
                    break;
                }
                case adaptivecards_controls_1.Constants.keys.left: {
                    _this.handleKeyboardLeftNav();
                    break;
                }
                case adaptivecards_controls_1.Constants.keys.right: {
                    _this.handleKeyboardRightNav();
                    break;
                }
                default: {
                    handled = false;
                }
            }
            if (handled) {
                e.preventDefault();
                e.cancelBubble = true;
            }
        };
        this._treeItemElement = document.createElement("div");
        this._treeItemElement.classList.add("acd-tree-item");
        this._treeItemElement.style.display = "flex";
        this._treeItemElement.style.alignItems = "center";
        this._treeItemElement.style.paddingLeft = this.getIndentationLevelIncrement() * (1 + this.level) + "px";
        this._treeItemElement.classList.toggle("selected", this._isSelected);
        this._treeItemElement.setAttribute("aria-labelledby", labelId);
        this._expandCollapseElement = document.createElement("div");
        this._expandCollapseElement.classList.add("acd-tree-item-expandCollapseButton");
        this._expandCollapseElement.style.flex = "0 0 auto";
        this._expandCollapseElement.style.visibility = hasChildren ? "visible" : "hidden";
        this._expandCollapseElement.setAttribute("role", "button");
        this._expandCollapseElement.setAttribute("aria-label", this.getExpandCollapseAriaText());
        this._expandCollapseElement.onclick = function (e) {
            _this._isExpanded = !_this._isExpanded;
            _this.updateLayout();
            e.cancelBubble = true;
            e.preventDefault();
        };
        this._treeItemElement.appendChild(this._expandCollapseElement);
        var textElement = document.createElement("div");
        textElement.classList.add("acd-tree-item-text");
        textElement.style.flex = "1 1 auto";
        textElement.style.display = "flex";
        textElement.style.alignItems = "center";
        textElement.style.whiteSpace = "nowrap";
        textElement.style.textOverflow = "ellipsis";
        textElement.style.overflow = "hidden";
        textElement.style.marginRight = "8px";
        if (this.getIconClass()) {
            var iconElement = document.createElement("div");
            iconElement.classList.add("acd-icon", "acd-treeView-icon", this.getIconClass());
            textElement.appendChild(iconElement);
        }
        var labelSpan = document.createElement("span");
        labelSpan.classList.add("acd-tree-item-typeName");
        labelSpan.innerText = this.getLabelText();
        labelSpan.id = labelId;
        textElement.appendChild(labelSpan);
        var text = this.getAdditionalText();
        if (text && text != "") {
            var additionalTextSpan = document.createElement("span");
            additionalTextSpan.classList.add(this.getAdditionalTextClass());
            additionalTextSpan.innerText = " [" + text + "]";
            textElement.appendChild(additionalTextSpan);
        }
        this._treeItemElement.appendChild(textElement);
        this._rootElement.appendChild(this._treeItemElement);
        if (hasChildren) {
            this._childContainerElement = document.createElement("ul");
            this._childContainerElement.style.paddingLeft = "0";
            this._childContainerElement.style.listStyleType = "none";
            this._childContainerElement.setAttribute("role", "group");
            for (var i = 0; i < this.getChildCount(); i++) {
                var renderedChildItem = this.getChildAt(i).render();
                renderedChildItem.setAttribute("aria-posinset", (i + 1).toString());
                renderedChildItem.setAttribute("aria-setsize", this.getChildCount().toString());
                this._childContainerElement.appendChild(renderedChildItem);
            }
            this._rootElement.appendChild(this._childContainerElement);
        }
        this.updateLayout();
        return this._rootElement;
    };
    BaseTreeItem.prototype.updateLayout = function () {
        var _a;
        if (this.getChildCount() > 0) {
            this._rootElement.setAttribute("aria-expanded", this._isExpanded.toString());
            this._expandCollapseElement.setAttribute("aria-label", this.getExpandCollapseAriaText());
            this._expandCollapseElement.classList.toggle(BaseTreeItem.expandedIconClass, this._isExpanded);
            this._expandCollapseElement.classList.toggle(BaseTreeItem.collapsedIconClass, !this._isExpanded);
            (_a = this._childContainerElement) === null || _a === void 0 ? void 0 : _a.classList.toggle("acd-hidden", !this._isExpanded);
        }
    };
    BaseTreeItem.prototype.expand = function () {
        if (!this._isExpanded) {
            this._isExpanded = true;
            this.updateLayout();
        }
    };
    BaseTreeItem.prototype.collpase = function () {
        if (this._isExpanded) {
            this._isExpanded = false;
            this.updateLayout();
        }
    };
    BaseTreeItem.prototype.isDraggable = function () {
        return false;
    };
    Object.defineProperty(BaseTreeItem.prototype, "level", {
        get: function () {
            return this._level;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseTreeItem.prototype, "isSelected", {
        get: function () {
            return this._isSelected;
        },
        set: function (value) {
            this.setIsSelected(value, true);
        },
        enumerable: false,
        configurable: true
    });
    BaseTreeItem.collapsedIconClass = "acd-icon-chevronRight";
    BaseTreeItem.expandedIconClass = "acd-icon-chevronDown";
    return BaseTreeItem;
}(draggable_element_1.DraggableElement));
exports.BaseTreeItem = BaseTreeItem;
